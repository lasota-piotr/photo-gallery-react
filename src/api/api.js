import unsplash from './unsplashMock';
import Unsplash, { toJson } from 'unsplash-js';
// import { unsplashData } from '../../data';
//
// const unsplash = new Unsplash(unsplashData);

export const listFeaturedCollections = () =>
  unsplash.collections.listFeaturedCollections(1, 4).then(toJson);

export const getCollection = (...args) =>
  unsplash.collections.getCollection(...args).then(toJson);

export const getCollectionPhotos = (...args) =>
  unsplash.collections.getCollectionPhotos(...args).then(toJson);

export const getPhoto = (...args) =>
  unsplash.photos.getPhoto(...args).then(toJson);
