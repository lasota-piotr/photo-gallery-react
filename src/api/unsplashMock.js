import {
  unsplashDataGetCollection,
  unsplashDatagetCollectionPhotos,
  unsplashDataListFeaturedCollections,
  unsplashDataGetPhotoWithCountry,
  unsplashDataGetPhotoWithoutCountry,
} from './unsplashMockData';
import delay from '../helpers/delay';

const unsplashMock = {
  collections: {
    listFeaturedCollections(_, num) {
      return delay(500).then(() =>
        unsplashDataListFeaturedCollections.slice(0, num)
      );
    },
    getCollection() {
      return delay(4000).then(() => unsplashDataGetCollection);
    },
    getCollectionPhotos(_, _1, num, order) {
      return delay(8000).then(() =>
        unsplashDatagetCollectionPhotos.slice(0, num)
      );
    },
  },
  photos: {
    getPhoto() {
      return delay(500).then(() => unsplashDataGetPhotoWithCountry);
    },
  },
};

export default unsplashMock;
