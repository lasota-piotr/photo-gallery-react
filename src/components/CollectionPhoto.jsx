import React from 'react';
import PropTypes from 'prop-types';
import { LOAD_STATE } from '../constants/constants';
import { getPhoto } from '../api/api';
import timeout from '../helpers/timout';
import CollectionPhotoContent from './CollectionPhotoContent';

class CollectionPhoto extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      loadStatePhotoInfo: LOAD_STATE.LOADING,
      photoInfo: {},
      mouseIsOver: false,
    };
    this.fetchPhotoInfo = this.fetchPhotoInfo.bind(this);
    this.onMouseEnterHandle = this.onMouseEnterHandle.bind(this);
    this.onMouseLeaveHandle = this.onMouseLeaveHandle.bind(this);
  }

  onMouseEnterHandle() {
    this.timeOutMouseEnter = timeout(500);
    this.timeOutMouseEnter.promise.then(() => {
      this.setState({ mouseIsOver: true });
      if (this.state.loadStatePhotoInfo !== LOAD_STATE.SUCCESS) {
        this.fetchPhotoInfo();
      }
    });
  }

  onMouseLeaveHandle() {
    if (this.timeOutMouseEnter && this.timeOutMouseEnter.cancel) {
      this.timeOutMouseEnter.cancel();
    }

    this.setState({ mouseIsOver: false });
  }

  fetchPhotoInfo() {
    const { id } = this.props.photo;
    getPhoto(id)
      .then(photoInfo => {
        this.setState({
          photoInfo,
          loadStatePhotoInfo: LOAD_STATE.SUCCESS,
        });
      })
      .catch(() => {
        this.setState({ loadStatePhotoInfo: LOAD_STATE.ERROR });
      });
  }

  render() {
    const { photo } = this.props;
    const { mouseIsOver, loadStatePhotoInfo, photoInfo } = this.state;
    return (
      <CollectionPhotoContent
        mouseIsOver={mouseIsOver}
        photoInfoIsLoading={loadStatePhotoInfo === LOAD_STATE.LOADING}
        photoInfoFromCollection={photo}
        photoInfo={photoInfo}
        onMouseEnterHandle={this.onMouseEnterHandle}
        onMouseLeaveHandle={this.onMouseLeaveHandle}
      />
    );
  }
}

CollectionPhoto.propTypes = {
  photo: PropTypes.shape({
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  }).isRequired,
};

export default CollectionPhoto;
