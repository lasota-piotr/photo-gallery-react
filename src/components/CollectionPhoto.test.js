import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import CollectionPhoto from './CollectionPhoto';
import { unsplashDatagetCollectionPhotos } from '../api/unsplashMockData';

jest.useFakeTimers();

test('Generates CollectionPhoto', () => {
  const component = renderer.create(
    <MemoryRouter>
      <CollectionPhoto photo={unsplashDatagetCollectionPhotos[0]} />
    </MemoryRouter>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  const innerDiv = tree.children[0].props;

  // manually trigger the callback
  innerDiv.onMouseEnter();
  expect(setTimeout).toHaveBeenCalledTimes(1);
  expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 500);

  // manually trigger the callback
  innerDiv.onMouseLeave();
  expect(setTimeout).toHaveBeenCalledTimes(1);
  expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 500);
});
