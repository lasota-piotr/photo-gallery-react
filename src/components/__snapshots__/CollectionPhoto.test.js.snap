// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generates CollectionPhoto 1`] = `
<a
  href="/photos/jMFo7nlYpUc"
  onClick={[Function]}
>
  <div
    aria-label=""
    className="sc-EHOje dhfkeg sc-bdVaJa iCgsfd"
    color="#fff"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    role="img"
    src="https://images.unsplash.com/photo-1516024802505-dca3acba2326?ixlib=rb-0.3.5&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=400&fit=max&s=a3712d1393d8ce3937fc3c5220f00987"
  />
</a>
`;
